File: tam_agent/analysis/reconcile.py
"""Reconcile top-down and bottom-up TAM estimates."""

from __future__ import annotations

from tam_agent.llm.client import LLMClient
from tam_agent.llm.prompts import RECONCILE_TAM, SYSTEM_PROMPT
from tam_agent.models import MarketEstimate


def reconcile_tam(
    llm: LLMClient,
    company_name: str,
    geography: str,
    top_down: MarketEstimate,
    bottom_up: MarketEstimate,
) -> MarketEstimate:
    """Produce a consensus TAM from top-down and bottom-up estimates."""

    prompt = RECONCILE_TAM.format(
        name=company_name,
        geography=geography,
        top_down_value=top_down.value_usd,
        top_down_confidence=top_down.confidence,
        top_down_method=top_down.method,
        top_down_assumptions="\n  - ".join(top_down.assumptions),
        bottom_up_value=bottom_up.value_usd,
        bottom_up_confidence=bottom_up.confidence,
        bottom_up_method=bottom_up.method,
        bottom_up_assumptions="\n  - ".join(bottom_up.assumptions),
    )

    return llm.structured_query(
        prompt=prompt,
        response_model=MarketEstimate,
        system=SYSTEM_PROMPT,
    )

File: src/analysis/reconcile.ts
/** Reconcile top-down and bottom-up TAM estimates. */

import type { LLMClient } from "../llm/client.js";
import { reconcileTamPrompt, SYSTEM_PROMPT } from "../llm/prompts.js";
import { MarketEstimateSchema, type MarketEstimate } from "../models.js";
import { formatUsd } from "../util/format.js";

export async function reconcileTam(
  llm: LLMClient,
  companyName: string,
  geography: string,
  topDown: MarketEstimate,
  bottomUp: MarketEstimate,
): Promise<MarketEstimate> {
  const prompt = reconcileTamPrompt({
    name: companyName,
    geography,
    top_down_value: formatUsd(topDown.value_usd),
    top_down_confidence: topDown.confidence,
    top_down_method: topDown.method,
    top_down_assumptions: topDown.assumptions.map((a) => `  - ${a}`).join("\n"),
    bottom_up_value: formatUsd(bottomUp.value_usd),
    bottom_up_confidence: bottomUp.confidence,
    bottom_up_method: bottomUp.method,
    bottom_up_assumptions: bottomUp.assumptions
      .map((a) => `  - ${a}`)
      .join("\n"),
  });

  return llm.structuredQuery(
    prompt,
    MarketEstimateSchema,
    "MarketEstimate",
    SYSTEM_PROMPT,
  );
}
